### PR TITLE
Added net7 support.

### PR DESCRIPTION
**Pull request by Haplois**

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -312,8 +312,8 @@ csharp_prefer_simple_default_expression = true:suggestion
 csharp_style_deconstructed_variable_declaration = true:suggestion
 csharp_style_inlined_variable_declaration = true:suggestion
 csharp_style_prefer_local_over_anonymous_function = false:suggestion # not default, default is true, avoided because we use anonymous functions in multiple places and it does not make the code clearer
-csharp_style_prefer_index_operator = true:suggestion
-csharp_style_prefer_range_operator = true:suggestion
+csharp_style_prefer_index_operator = false:suggestion # not default, default is true, disabled because Index operator is not supported in all the TFMs we support.
```
Comment (nohwnd) [inline .editorconfig:315]: Do you know which tfms don't support it? It would be worth mentioning it here so we don't have to investigate the motifvation for this comment when we move on to next version of netfx.

File: src/Microsoft.TestPlatform.CoreUtilities/PublicAPI/net7.0/PublicAPI.Shipped.txt [hunk 2]
```diff
@@ -0,0 +1,13 @@
+﻿#nullable enable
```
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CoreUtilities/PublicAPI/net7.0/PublicAPI.Shipped.txt:1]: Is this file identical to the one for net6.0? Can we just rename the other file and include the same file for both tfms in the csproj? I don't think these files will grow apart in the future.

Comment (nohwnd) [reply src/Microsoft.TestPlatform.CoreUtilities/PublicAPI/net7.0/PublicAPI.Shipped.txt:1]: https://github.com/microsoft/vstest/blob/main/src/Microsoft.TestPlatform.CoreUtilities/PublicAPI/net6.0/PublicAPI.Shipped.txt They seem identical on first look.

File: src/Microsoft.TestPlatform.CoreUtilities/ValidateArg.cs [hunk 3]
```diff
@@ -34,7 +34,7 @@ public static class ValidateArg
     /// Type of argument.
     /// </returns>
     [DebuggerStepThrough]
-    public static T NotNull<T>([ValidatedNotNull] T? arg, string parameterName)
+    public static T NotNull<T>([NotNull] T? arg, string parameterName)
```
Comment (nohwnd) [inline src/Microsoft.TestPlatform.CoreUtilities/ValidateArg.cs:37]: What is the impact of this change?

File: src/Microsoft.TestPlatform.ObjectModel/PublicAPI/net7.0/PublicAPI.Shipped.txt [hunk 4]
```diff
@@ -0,0 +1,4 @@
+﻿#nullable enable
```
Comment (nohwnd) [inline src/Microsoft.TestPlatform.ObjectModel/PublicAPI/net7.0/PublicAPI.Shipped.txt:1]: Same question as above, is this just a copy of net6.0 file?

File: src/Microsoft.TestPlatform.PlatformAbstractions/PublicAPI/net7.0/PublicAPI.Shipped.txt [hunk 5]
```diff
@@ -0,0 +1,17 @@
+﻿#nullable enable
```
Comment (nohwnd) [inline src/Microsoft.TestPlatform.PlatformAbstractions/PublicAPI/net7.0/PublicAPI.Shipped.txt:1]: And here as well. Can this net6.0 file be reused here?

File: src/Microsoft.TestPlatform.PlatformAbstractions/netstandard/System/ProcessHelper.cs [hunk 6]
```diff
@@ -93,7 +93,7 @@ public void WaitForProcessExit(object? process)
         throw new NotImplementedException();
     }
 
-    public IntPtr GetProcessHandle(int processId)
+    public nint GetProcessHandle(int processId)
```
Comment (nohwnd) [inline src/Microsoft.TestPlatform.PlatformAbstractions/netstandard/System/ProcessHelper.cs:96]: did not even know that nint exists, cool.

File: src/Microsoft.TestPlatform.CoreUtilities/ValidateArg.cs (hunk 3 above)
Comment (Haplois) [reply src/Microsoft.TestPlatform.CoreUtilities/ValidateArg.cs:37]: No impact. That also works on net7.

File: .editorconfig (hunk 1 above)
Comment (Haplois) [reply .editorconfig:315]: Added supported versions.

File: src/Microsoft.TestPlatform.ObjectModel/PublicAPI/net7.0/PublicAPI.Shipped.txt (hunk 4 above)
Comment (Haplois) [reply src/Microsoft.TestPlatform.ObjectModel/PublicAPI/net7.0/PublicAPI.Shipped.txt:1]: no net6 here.

File: src/DataCollectors/Microsoft.TestPlatform.Extensions.EventLogCollector/EventLogDataCollector.cs [hunk 7]
```diff
@@ -566,16 +566,17 @@ private void ConfigureMaxEntries(CollectorNameValueConfigurationManager collecto
...
-        EventLogSessionContext eventLogSessionContext;
+        EventLogSessionContext? eventLogSessionContext;
         bool eventLogContainerFound;
         lock (ContextMap)
         {
             eventLogContainerFound = ContextMap.TryGetValue(dataCollectionContext, out eventLogSessionContext);
         }
 
-        if (!eventLogContainerFound)
+        if (eventLogContainerFound == false)
```
Comment (Evangelink) [inline src/DataCollectors/Microsoft.TestPlatform.Extensions.EventLogCollector/EventLogDataCollector.cs:579]: Previous code was ok to me. Usually when I see `== false`, I expect some nullable bool which isn't the case here.

File: src/DataCollectors/Microsoft.TestPlatform.Extensions.EventLogCollector/Microsoft.TestPlatform.Extensions.EventLogCollector.csproj [hunk 8]
```diff
@@ -6,18 +6,12 @@
...
     <UseBannedApiAnalyzers>true</UseBannedApiAnalyzers>
   </PropertyGroup>
-  <ItemGroup Condition=" '$(DotNetBuildFromSource)' != 'true' ">
-    <Reference Include="System" />
-    <Reference Include="System.Runtime" />
-    <Reference Include="Microsoft.CSharp" />
-    <Reference Include="System.Xml" />
-  </ItemGroup>
-  <ItemGroup Condition="'$(TargetFramework)' == 'net6.0'">
+  <ItemGroup Condition="'$(TargetFramework)' == 'net7.0'">
```
Comment (Evangelink) [inline src/DataCollectors/Microsoft.TestPlatform.Extensions.EventLogCollector/Microsoft.TestPlatform.Extensions.EventLogCollector.csproj:14]: Could we use `'$(DotNetBuildFromSource)' == 'true'` instead? This would avoid us to update the TFM every time.

File: src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionAttachmentManager.cs [hunk 9]
```diff
@@ -290,7 +290,12 @@ private void AddNewFileTransfer(FileTransferInformation fileTransferInfo, AsyncC
                         ex.ToString(),
                         uri,
                         friendlyName,
-                        Guid.Parse(testCaseId));
+#if NET7_0_OR_GREATER
+                        Guid.Parse(testCaseId, CultureInfo.InvariantCulture)
+#else
+                        Guid.Parse(testCaseId)
+#endif
```
Comment (Evangelink) [inline src/Microsoft.TestPlatform.Common/DataCollection/DataCollectionAttachmentManager.cs:297]: Not sure if that's way better but an alternative is

```suggestion
                        Guid.Parse(testCaseId
#if NET7_0_OR_GREATER
                            , CultureInfo.InvariantCulture
#endif
                            )
```

File: src/Microsoft.TestPlatform.Extensions.BlameDataCollector/WindowsHangDumper.cs [hunk 10]
```diff
@@ -131,17 +131,17 @@ internal static void CollectDump(IProcessHelper processHelper, Process process,
                 PlatformArchitecture.X86 => "DumpMinitool.x86.exe",
                 PlatformArchitecture.X64 => "DumpMinitool.exe",
                 PlatformArchitecture.ARM64 => "DumpMinitool.arm64.exe",
-                _ => null
+                _ => string.Empty
             };
 
-            if (dumpMinitoolName == null)
+            if (dumpMinitoolName == string.Empty)
```
Comment (Evangelink) [inline src/Microsoft.TestPlatform.Extensions.BlameDataCollector/WindowsHangDumper.cs:137]: Why this change from `null` to string empty?

File: src/Microsoft.TestPlatform.ObjectModel/TestCase.cs [hunk 11]
```diff
@@ -261,7 +262,13 @@ protected override void ProtectedSetPropertyValue(TestProperty property, object?
                 return;
 
             case "TestCase.Id":
-                Id = value is Guid guid ? guid : Guid.Parse((value as string)!);
+                Id = value is Guid guid
+                    ? guid
+#if NET7_0_OR_GREATER
+                    : Guid.Parse((value as string)!, CultureInfo.InvariantCulture);
```
Comment (Evangelink) [inline src/Microsoft.TestPlatform.ObjectModel/TestCase.cs:268]: Ouch there is something fishy here. We should not be doing `as` cast to then bang, it's either nullable or we should use direct cast instead.

File: src/Microsoft.TestPlatform.PlatformAbstractions/PublicAPI/PublicAPI.Shipped.txt [hunk 12]
```diff
@@ -47,7 +47,7 @@ Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces.IProcessHelp
 Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces.IProcessHelper.GetCurrentProcessId() -> int
 Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces.IProcessHelper.GetCurrentProcessLocation() -> string!
 Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces.IProcessHelper.GetNativeDllDirectory() -> string!
-Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces.IProcessHelper.GetProcessHandle(int processId) -> System.IntPtr
+Microsoft.VisualStudio.TestPlatform.PlatformAbstractions.Interfaces.IProcessHelper.GetProcessHandle(int processId) -> nint
```
Comment (Evangelink) [inline src/Microsoft.TestPlatform.PlatformAbstractions/PublicAPI/PublicAPI.Shipped.txt:50]: Isn't this going to break compat for old TFMs?

File: src/Microsoft.TestPlatform.Utilities/ClientUtilities.cs [hunk 13]
```diff
@@ -54,6 +54,8 @@ public static void FixRelativePathsInRunSettings(XmlDocument xmlDocument, string
 
     private static void AddRunSettingsDirectoryNode(XmlDocument doc, string path)
     {
+        ValidateArg.NotNull(doc.DocumentElement, nameof(doc.DocumentElement));
```
Comment (Evangelink) [inline src/Microsoft.TestPlatform.Utilities/ClientUtilities.cs:57]: I'm not a big fan of this. This is going to throw `ArgumentNullException` while it's not the arg that is null but a field of it. IMO we should throw either `ArgumentException` or `InvalidOperationException` (I would go for the last one as I mentally associate it to "invalid state" but the first one is fine too).

File: src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs [hunk 14]
```diff
@@ -705,12 +705,8 @@ public static IEnumerable<string> FilterCompatibleSources(Architecture chosenPla
...
     private static bool IsPlatformIncompatible(Architecture sourcePlatform, Architecture targetPlatform)
     {
-        if (sourcePlatform is Architecture.Default or Architecture.AnyCPU)
-        {
-            return false;
-        }
-
-        return targetPlatform == Architecture.X64 && !Is64BitOperatingSystem() || sourcePlatform != targetPlatform;
+        return sourcePlatform is not Architecture.Default and not Architecture.AnyCPU
+&& (targetPlatform == Architecture.X64 && !Is64BitOperatingSystem() || sourcePlatform != targetPlatform);
```
Comment (Evangelink) [inline src/Microsoft.TestPlatform.Utilities/InferRunSettingsHelper.cs:709]: Nit: Indentation is broken here.

File: src/Microsoft.TestPlatform.Utilities/XmlUtilities.cs [hunk 15]
```diff
@@ -63,7 +63,7 @@ internal static bool IsValidNodeXmlValue(string xmlNodeValue, Func<string, bool>
         var secureInnerXml = SecurityElement.Escape(innerXml);
 #else
         // fixing manually as we currently target to netcore 1.1 and we don't have default implementation for Escape functionality
-        var secureInnerXml = innerXml.IsNullOrEmpty() ? innerXml : innerXml.Replace("&", "&amp;").Replace("<", "&lt;").Replace(">", "&gt;").Replace("\"", "&quot;").Replace("'", "&apos;");
+        var secureInnerXml = (innerXml ?? string.Empty).Replace("&", "&amp;").Replace("<", "&lt;").Replace(">", "&gt;").Replace("\"", "&quot;").Replace("'", "&apos;");
```
Comment (Evangelink) [inline src/Microsoft.TestPlatform.Utilities/XmlUtilities.cs:66]: You have changed the logic here. That's not a big deal but you are forcing unnecessary operations

File: src/SettingsMigrator/Migrator.cs [hunk 16]
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
     /// <param name="runSettingsXmlDoc">Runsettings Xml</param>
     private static void MigrateTestSettingsNodesToRunSettings(string testSettingsPath, XmlDocument runSettingsXmlDoc)
     {
-        var testSettingsNodes = ReadTestSettingsNodes(testSettingsPath);
+        TestSettingsNodes testSettingsNodes = ReadTestSettingsNodes(testSettingsPath);
 
         string? testTimeout = null;
-        if (testSettingsNodes.Timeout != null && testSettingsNodes.Timeout.Attributes[TestTimeoutAttributeName] != null)
+        if (testSettingsNodes?.Timeout?.Attributes?[TestTimeoutAttributeName] != null)
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:158]: According to nullability above, `testSettingsNodes`  cannot be null

```suggestion
        if (testSettingsNodes.Timeout?.Attributes?[TestTimeoutAttributeName] != null)
```

File: src/SettingsMigrator/Migrator.cs [hunk 17]
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
...
         {
-            testTimeout = testSettingsNodes.Timeout.Attributes[TestTimeoutAttributeName].Value;
+            testTimeout = testSettingsNodes?.Timeout?.Attributes?[TestTimeoutAttributeName]?.Value;
         }
 
         string? runTimeout = null;
-        if (testSettingsNodes.Timeout != null && testSettingsNodes.Timeout.Attributes[RunTimeoutAttributeName] != null)
+        if (testSettingsNodes?.Timeout?.Attributes?[RunTimeoutAttributeName] != null)
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:164]: According to nullability above, `testSettingsNodes`  cannot be null

```suggestion
        if (testSettingsNodes.Timeout?.Attributes?[RunTimeoutAttributeName] != null)
```

File: src/SettingsMigrator/Migrator.cs [hunk 18]
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
...
         {
-            runTimeout = testSettingsNodes.Timeout.Attributes[RunTimeoutAttributeName].Value;
+            runTimeout = testSettingsNodes?.Timeout?.Attributes?[RunTimeoutAttributeName]?.Value;
         }
 
         string? parallelTestCount = null;
-        if (testSettingsNodes.Execution != null && testSettingsNodes.Execution.Attributes[ParallelTestCountAttributeName] != null)
+        if (testSettingsNodes?.Execution?.Attributes?[ParallelTestCountAttributeName] != null)
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:170]: According to nullability above, `testSettingsNodes`  cannot be null

```suggestion
        if (testSettingsNodes.Execution?.Attributes?[ParallelTestCountAttributeName] != null)
```

File: src/SettingsMigrator/Migrator.cs [hunk 19]
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
...
         {
-            parallelTestCount = testSettingsNodes.Execution.Attributes[ParallelTestCountAttributeName].Value;
+            parallelTestCount = testSettingsNodes?.Execution?.Attributes?[ParallelTestCountAttributeName]?.Value;
         }
 
         string? hostProcessPlatform = null;
-        if (testSettingsNodes.Execution != null && testSettingsNodes.Execution.Attributes[HostProcessPlatformAttributeName] != null)
+        if (testSettingsNodes?.Execution?.Attributes?[HostProcessPlatformAttributeName] != null)
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:176]: According to nullability above, `testSettingsNodes`  cannot be null

```suggestion
        if (testSettingsNodes.Execution?.Attributes?[HostProcessPlatformAttributeName] != null)
```

File: src/SettingsMigrator/Migrator.cs [hunk 20]
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
...
         {
-            hostProcessPlatform = testSettingsNodes.Execution.Attributes[HostProcessPlatformAttributeName].Value;
+            hostProcessPlatform = testSettingsNodes?.Execution?.Attributes?[HostProcessPlatformAttributeName]?.Value;
         }
 
         // WebTestRunConfiguration node.
-        if (testSettingsNodes.WebSettings != null)
+        if (testSettingsNodes?.WebSettings != null)
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:182]: According to nullability above, `testSettingsNodes`  cannot be null

```suggestion
        if (testSettingsNodes.WebSettings != null)
```

File: src/SettingsMigrator/Migrator.cs [hunk 21]
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
...
         {
-            runSettingsXmlDoc.DocumentElement.AppendChild(runSettingsXmlDoc.ImportNode(testSettingsNodes.WebSettings, deep: true));
+            runSettingsXmlDoc?.DocumentElement?.AppendChild(runSettingsXmlDoc.ImportNode(testSettingsNodes.WebSettings, deep: true));
         }
 
         // LegacySettings node.
-        AddLegacyNodes(testSettingsNodes, testTimeout, parallelTestCount, hostProcessPlatform, runSettingsXmlDoc);
+        AddLegacyNodes(testSettingsNodes!, testTimeout, parallelTestCount, hostProcessPlatform, runSettingsXmlDoc!);
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:188]: According to nullability above, `testSettingsNodes`  cannot be null

```suggestion
        AddLegacyNodes(testSettingsNodes, testTimeout, parallelTestCount, hostProcessPlatform, runSettingsXmlDoc!);
```

File: src/SettingsMigrator/Migrator.cs [hunk 22]
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
...
         // TestSessionTimeout node.
         if (!runTimeout.IsNullOrEmpty())
         {
-            AddRunTimeoutNode(runTimeout, runSettingsXmlDoc);
+            AddRunTimeoutNode(runTimeout, runSettingsXmlDoc!);
         }
 
         // DataCollectors node.
-        if (testSettingsNodes.Datacollectors != null && testSettingsNodes.Datacollectors.Count > 0)
+        if (testSettingsNodes?.Datacollectors != null && testSettingsNodes.Datacollectors.Count > 0)
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:197]: According to nullability above, `testSettingsNodes`  cannot be null

```suggestion
        if (testSettingsNodes.Datacollectors != null && testSettingsNodes.Datacollectors.Count > 0)
```

File: src/SettingsMigrator/Migrator.cs [hunk 23]
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
...
-            testTimeout = testSettingsNodes.Timeout.Attributes[TestTimeoutAttributeName].Value;
+            testTimeout = testSettingsNodes?.Timeout?.Attributes?[TestTimeoutAttributeName]?.Value;
         }
 
         string? runTimeout = null;
-        if (testSettingsNodes.Timeout != null && testSettingsNodes.Timeout.Attributes[RunTimeoutAttributeName] != null)
+        if (testSettingsNodes?.Timeout?.Attributes?[RunTimeoutAttributeName] != null)
         {
-            runTimeout = testSettingsNodes.Timeout.Attributes[RunTimeoutAttributeName].Value;
+            runTimeout = testSettingsNodes?.Timeout?.Attributes?[RunTimeoutAttributeName]?.Value;
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:166]: You checked for null above so you don't need the various conditional access.

File: src/SettingsMigrator/Migrator.cs [hunk 24]
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
...
         }
 
         string? parallelTestCount = null;
-        if (testSettingsNodes.Execution != null && testSettingsNodes.Execution.Attributes[ParallelTestCountAttributeName] != null)
+        if (testSettingsNodes?.Execution?.Attributes?[ParallelTestCountAttributeName] != null)
         {
-            parallelTestCount = testSettingsNodes.Execution.Attributes[ParallelTestCountAttributeName].Value;
+            parallelTestCount = testSettingsNodes?.Execution?.Attributes?[ParallelTestCountAttributeName]?.Value;
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:172]: You checked for null above so you don't need the various conditional access.

File: src/SettingsMigrator/Migrator.cs [hunk 25]
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
...
         }
 
         string? hostProcessPlatform = null;
-        if (testSettingsNodes.Execution != null && testSettingsNodes.Execution.Attributes[HostProcessPlatformAttributeName] != null)
+        if (testSettingsNodes?.Execution?.Attributes?[HostProcessPlatformAttributeName] != null)
         {
-            hostProcessPlatform = testSettingsNodes.Execution.Attributes[HostProcessPlatformAttributeName].Value;
+            hostProcessPlatform = testSettingsNodes?.Execution?.Attributes?[HostProcessPlatformAttributeName]?.Value;
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:178]: You checked for null above so you don't need the various conditional access.

File: src/SettingsMigrator/Migrator.cs [hunk 26]
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
...
         }
 
         // WebTestRunConfiguration node.
-        if (testSettingsNodes.WebSettings != null)
+        if (testSettingsNodes?.WebSettings != null)
         {
-            runSettingsXmlDoc.DocumentElement.AppendChild(runSettingsXmlDoc.ImportNode(testSettingsNodes.WebSettings, deep: true));
+            runSettingsXmlDoc?.DocumentElement?.AppendChild(runSettingsXmlDoc.ImportNode(testSettingsNodes.WebSettings, deep: true));
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:184]: `runSettingsXmlDoc` cannot be null, please remove the conditional access.

File: src/SettingsMigrator/Migrator.cs [hunk 27]
```diff
@@ -153,65 +152,65 @@ private static void MigrateTestSettings(string oldTestSettingsPath, string newRu
...
 
         // LegacySettings node.
-        AddLegacyNodes(testSettingsNodes, testTimeout, parallelTestCount, hostProcessPlatform, runSettingsXmlDoc);
+        AddLegacyNodes(testSettingsNodes!, testTimeout, parallelTestCount, hostProcessPlatform, runSettingsXmlDoc!);
 
         // TestSessionTimeout node.
         if (!runTimeout.IsNullOrEmpty())
         {
-            AddRunTimeoutNode(runTimeout, runSettingsXmlDoc);
+            AddRunTimeoutNode(runTimeout, runSettingsXmlDoc!);
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:193]: Cannot be null

```suggestion
            AddRunTimeoutNode(runTimeout, runSettingsXmlDoc);
```

File: src/SettingsMigrator/Migrator.cs (hunk 22 above)
Comment (Evangelink) [reply src/SettingsMigrator/Migrator.cs:197]: I believe you can also simplify further

```suggestion
        if (testSettingsNodes.Datacollectors?.Count > 0)
```

File: src/SettingsMigrator/Migrator.cs [hunk 28]
```diff
@@ -223,14 +222,14 @@ private static TestSettingsNodes ReadTestSettingsNodes(string testSettingsPath)
             testSettingsNodes.Hosts = testSettingsRoot.SelectSingleNode(@"/TestSettings/Execution/Hosts");
             testSettingsNodes.Execution = testSettingsRoot.SelectSingleNode(@"/TestSettings/Execution");
 
-            if (testSettingsNodes.Timeout != null && (testSettingsNodes.Timeout.Attributes[AgentNotRespondingTimeoutAttribute] != null ||
-                                                      testSettingsNodes.Timeout.Attributes[DeploymentTimeoutAttribute] != null || testSettingsNodes.Timeout.Attributes[ScriptTimeoutAttribute] != null))
+            if (testSettingsNodes.Timeout != null && (testSettingsNodes?.Timeout?.Attributes?[AgentNotRespondingTimeoutAttribute] != null ||
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:225]: `testSettingsXmlDoc` cannot be null.

File: src/SettingsMigrator/Migrator.cs [hunk 29]
```diff
@@ -223,14 +222,14 @@ private static TestSettingsNodes ReadTestSettingsNodes(string testSettingsPath)
...
+                                                      testSettingsNodes?.Timeout?.Attributes?[DeploymentTimeoutAttribute] != null || testSettingsNodes?.Timeout?.Attributes?[ScriptTimeoutAttribute] != null))
             {
                 Console.WriteLine(CommandLineResources.UnsupportedAttributes);
             }
         }
 
-        return testSettingsNodes;
+        return testSettingsNodes!;
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:232]: Cannot be null.

```suggestion
        return testSettingsNodes;
```

File: src/SettingsMigrator/Migrator.cs [hunk 30]
```diff
@@ -239,10 +238,10 @@ private static TestSettingsNodes ReadTestSettingsNodes(string testSettingsPath)
     /// <param name="newXmlDoc">Xml doc to process</param>
     private static void RemoveEmbeddedTestSettings(XmlDocument newXmlDoc)
     {
-        var testSettingsNode = newXmlDoc.DocumentElement.SelectSingleNode(@"/RunSettings/MSTest/SettingsFile");
+        var testSettingsNode = newXmlDoc?.DocumentElement?.SelectSingleNode(@"/RunSettings/MSTest/SettingsFile");
```
Comment (Evangelink) [inline src/SettingsMigrator/Migrator.cs:241]: `newXmlDoc` cannot be null.

Comment (Evangelink) [reply src/SettingsMigrator/Migrator.cs:241]: Can you please review all nullability in this class.

File: src/vstest.console/CommandLine/AssemblyMetadataProvider.cs [hunk 31]
```diff
@@ -58,7 +58,15 @@ public Architecture GetArchitecture(string assemblyPath)
         {
             // AssemblyName won't load the assembly into current domain.
             var assemblyName = AssemblyName.GetAssemblyName(assemblyPath);
-            archType = MapToArchitecture(assemblyName.ProcessorArchitecture, assemblyPath);
+
+            var processorArchitecture =
```
Comment (Evangelink) [inline src/vstest.console/CommandLine/AssemblyMetadataProvider.cs:62]: I believe this deserves a code explanation.

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs [hunk 32]
```diff
@@ -81,7 +80,12 @@ public InProcessVsTestConsoleWrapper(ConsoleParameters consoleParameters)
         }
 
         // Fill the parameters.
-        consoleParameters.ParentProcessId = Process.GetCurrentProcess().Id;
+        consoleParameters.ParentProcessId =
+#if NET7_0_OR_GREATER
```
Comment (Evangelink) [inline src/vstest.console/InProcessVsTestConsoleWrapper.cs:84]: I should have called this earlier but can you please add a helper for this so that the pragma is only in one place and all calling sites have a simpler code?

File: src/Microsoft.TestPlatform.CoreUtilities/ValidateArg.cs (hunk 3 above)
Comment (nohwnd) [reply src/Microsoft.TestPlatform.CoreUtilities/ValidateArg.cs:37]: I was looking for motivation of this change. Why is this better than the attribute we had?

File: shared/NullableAttributes.cs [hunk 33]
```diff
@@ -149,11 +146,3 @@ public MemberNotNullWhenAttribute(bool returnValue, params string[] members)
 }
 
 #endif
-
-/// <summary>
-/// Secret attribute that tells the CA1062 validate arguments rule that this method validates the argument is not null.
-/// </summary>
-[AttributeUsage(AttributeTargets.Parameter)]
-internal sealed class ValidatedNotNullAttribute : Attribute
```
[the remark this answers is not in the document]
Comment (nohwnd) [reply shared/NullableAttributes.cs:157]: Asked similar thing above. https://github.com/microsoft/vstest/pull/3944#discussion_r951173250

File: src/Microsoft.TestPlatform.CoreUtilities/ValidateArg.cs (hunk 3 above)
Comment (Evangelink) [reply src/Microsoft.TestPlatform.CoreUtilities/ValidateArg.cs:37]: It's shipped inside the framework starting with .net 5 or 6 while the other one has to be manually defined.

Comment (Haplois) [reply src/Microsoft.TestPlatform.CoreUtilities/ValidateArg.cs:37]: Attribute we had didn't suppress the warnings and we already had this so in the cases it suppressed it was just a duplicate.

File: shared/NullableAttributes.cs (hunk 33 above)
[the remark this answers is not in the document]
Comment (Evangelink) [reply shared/NullableAttributes.cs:157]: Closing this one as I have replied on the other one. Looking at the usage gave me the answer.

File: src/DataCollectors/Microsoft.TestPlatform.Extensions.EventLogCollector/EventLogDataCollector.cs [hunk 34]
```diff
@@ -566,16 +566,17 @@ private void ConfigureMaxEntries(CollectorNameValueConfigurationManager collecto
         }
     }
 
+    [return: NotNull]
```
[the remark this answers is not in the document]
Comment (Haplois) [reply src/DataCollectors/Microsoft.TestPlatform.Extensions.EventLogCollector/EventLogDataCollector.cs:569]: It sees the return value as "might be null" without this.

File: src/DataCollectors/Microsoft.TestPlatform.Extensions.EventLogCollector/Microsoft.TestPlatform.Extensions.EventLogCollector.csproj (hunk 8 above)
Comment (Haplois) [reply src/DataCollectors/Microsoft.TestPlatform.Extensions.EventLogCollector/Microsoft.TestPlatform.Extensions.EventLogCollector.csproj:14]: No, not now. In a separate PR, I will remove most of the sourcebuild customizations and use the same targets. That change is a bit bigger, so duplicates for now.

File: src/Microsoft.TestPlatform.PlatformAbstractions/PublicAPI/PublicAPI.Shipped.txt (hunk 12 above)
Comment (Haplois) [reply src/Microsoft.TestPlatform.PlatformAbstractions/PublicAPI/PublicAPI.Shipped.txt:50]: No, as long as we are using a recent compiler.

File: playground/MSTest1/MSTest1.csproj [hunk 35]
```diff
@@ -6,7 +6,7 @@
   <Import Project="$(TestPlatformRoot)scripts/build/TestPlatform.Settings.targets" />
 
   <PropertyGroup>
-    <TargetFrameworks>$(TargetFrameworks);net472;net5.0</TargetFrameworks>
+    <TargetFrameworks>$(TargetFrameworks);net472;net5.0;net6.0;net7.0</TargetFrameworks>
```
[the remark this answers is not in the document]
Comment (Haplois) [reply playground/MSTest1/MSTest1.csproj:9]: Not really, we can remove it. I will.

[the remark this answers is not in the document]
Comment (Haplois) [reply playground/MSTest1/MSTest1.csproj:9]: No, not now. In a separate PR, I will remove most of the sourcebuild customizations and use the same targets. That change is a bit bigger, so duplicates for now.

File: src/SettingsMigrator/Migrator.cs (hunk 23 above)
Comment (Haplois) [reply src/SettingsMigrator/Migrator.cs:166]: I do, it causes might be null warnings.

File: src/SettingsMigrator/Migrator.cs (hunk 24 above)
Comment (Haplois) [reply src/SettingsMigrator/Migrator.cs:172]: same, if I don't - causes might be null warnings.

File: src/SettingsMigrator/Migrator.cs (hunk 30 above)
Comment (Haplois) [reply src/SettingsMigrator/Migrator.cs:241]: I did. If I remove marks causes might be null warnings.

File: src/vstest.console/InProcessVsTestConsoleWrapper.cs (hunk 32 above)
Comment (Haplois) [reply src/vstest.console/InProcessVsTestConsoleWrapper.cs:84]: This change requires a lot of additional changes, (aka consolidating our fill-ins, making some internals visible etc. - we can do this, but in a different pr.)

**Review by Evangelink — APPROVED** [review]
Let's wait for tomorrow's discussion on netstandard before merging this one. There will be conflicts between the 2 PRs and netstandard is older so it would be better to fix conflicts here. If we decide to make more work on netstandard PR then we will merge this one first.

**Comment by Evangelink** [comment]
@Haplois Ping me for a new review when conflicts are fixed.

**Review by Evangelink — APPROVED** [review]
